### PR TITLE
Refactor: moving job initialization work into the client package

### PR DESCRIPTION
**Pull request by izaaklauer**
### What changed

This PR primarily moves the local runner trigger logic out of the CLI and into the project client. Rather than rely on each command to declare ahead of time if they require a runner, and start the runner before the command can run, now we wait until a CLI command actually attempts to spawn a job, and start the runner there.

Upsides to doing this:
- CLI commands don't need to explicitly declare if they need a runner. This is a big win for the `status` command, which only needs a runner conditionally (if `-refresh` is present). 
- The client package seems like the "right" place for this logic to live - as the primary internal interface to the waypoint APIs, it seems like it should be responsible for abstracting these details. With this change, all of the logic around handling the local runner lives inside the client package.

Downsides to doing this:
- The details of `logs` and `exec` _within_ the client package get a little bit messier. Unlike every other command, `logs` and `exec` don't create jobs from the CLI side. Instead, they need to start a local runner, call a server API, and then the server creates the job and decides if it needs to be assigned back to the local runner. So for logs and exec, we need to become ready to accept local jobs, but _not_ create a job. To make that happen, I've split all of the job spoolup logic into a function called `setupLocalJobSystem`, which is normally called within `doJob` but is _also_ called by the logs and exec client operation funcs.

This also sets the stage for us to add more complexity into the `setupLocalJobSystem` sequence - namely checking if VCS is dirty and issuing a warning or forwarding the `waypoint.hcl` file along with the job.

### Smaller incidental changes:
- Adds the operation set flags to logs and exec. They both could trigger operations, and without them you have no control over if those operations happen locally or remotely. As a result, up until now, it hasn't been possible to run exec or logs non-locally! I tested them both, and they work with remotely ODR.
- Wires in `remoteOpPreferred`, initially introduced in https://github.com/hashicorp/waypoint/pull/2749. Operations will now occur remotely if they're likely to work remotely, unless `-local` is specified.
- Misc typo fixes and improved logging

### CODE REVIEW

**Comment by mitchellh** [comment]
Historical fact: we used to do this, if you go back in git history this was how it originally worked. It was switched to CLI global at some point by @evanphx for some good reason but I don't exactly recall why. :)

File: internal/cli/status.go
Comment (izaaklauer) [inline internal/cli/status.go:137]: This was my primary impetus for moving the runner startup logic. Once we know we're going to need to run a job, we need to do a bunch of work:

- Figure out if we should run locally or remotely (based on flags, the project, and the waypoint infra's current state)
- If local, start the local runner
- If remote, check if VCS is dirty, and if so warn

Basically all the work that currently is in `setupLocalJobSystem`. If we don't move that logic into the `doJob` path, we'll need to do it here, which feels messier to me than making the logs and exec operation funcs call `setupLocalJobSystem`.

File: internal/client/job.go
Comment (izaaklauer) [inline internal/client/job.go:48]: Extra scrutiny here please. This whole function smells, but we're going to need to execute this bundle of logic _somewhere_ - either in the status CLI command (if we don't move the runner startup logic), or in the logs and exec operations (as this pr proposes). This is the least-worst option I've come up with, but i'm certainly open to better ideas.

File: internal/client/operation.go
Comment (izaaklauer) [inline internal/client/operation.go:163]: This function was unused previously. Now, the exec cli calls it to trigger the op. This gives the client package here the opportunity to call the internal `setupLocalJobSystem` method to spool up a local runner that can receive the job if the server decides it's necessary.

File: internal/client/runner.go
Comment (izaaklauer) [inline internal/client/runner.go:60]: This isn't new - just moved from the CLI's `base.go`.

**Comment by izaaklauer** [comment]
> Historical fact: we used to do this, if you go back in git history this was how it originally worked. It was switched to CLI global at some point by @evanphx for some good reason but I don't exactly recall why. :)

Thank you! I spoke to @evanphx about this, and I think there were two reasons it was moved historically:
- To make it so we only start one local runner per cli invocation, and re-use it if we perform multiple operations
- To make `logs` and `exec` work, which require a local runner but don't create a job in the CLI - the server creates one and decides to assign it back to the local CLI runner if necessary.

This change preserves both of those properties. That said, I'm not wedded to this refactor, and if there are other reasons that I'm not seeing to keep the runner startup logic cli-global, we don't need to go down this route.

File: internal/client/job.go
Comment (mitchellh) [inline internal/client/job.go:56]: Since you're using named return values ANYWAYS, I might suggest adding a defer to always log the results of this function:

```go
defer func() {
  c.log.Named("setupLocalJobSystem").Info("result", "isLocal", ..., etc.)
}()
```

File: internal/client/job.go
Comment (mitchellh) [inline internal/client/job.go:58]: Would love a comment above here reminding me what this tri-state being `nil` implies. i.e. "nil means we should auto-detect" or something.

File: internal/client/job.go
Comment (mitchellh) [inline internal/client/job.go:60]: I would add logging in this block using the logger on `c`

File: internal/client/job.go
Comment (mitchellh) [inline internal/client/job.go:54]: I think this function is sufficiently complicated and important that it may be worth unit testing in isolation.

**Review by mitchellh — CHANGES_REQUESTED** [review]
Overall, excellent, and I raelly like this location for this a LOT better.

**Comment by krantzinator** [comment]
Regarding this
>Adds the operation set flags to logs and exec. They both could trigger operations, and without them you have no control over if those operations happen locally or remotely. As a result, up until now, it hasn't been possible to run exec or logs non-locally! I tested them both, and they work with remotely ODR.
-- what platforms did you test this on? I assume k8s? I may test against nomad and/or ECS just to see if any oddities pop up.

**Comment by izaaklauer** [comment]
@krantzinator I tested on docker, k8s, and lambda - I picked lambda because it's the only plugin (I think) that has plugin-custom implementations for `logs` and `exec` - no other platform executes any plugin-specific code to service `logs` or `exec` (because they have a long-running ceb available to handle that). Nomad/ECS specific testing would be awesome though!

**Review by krantzinator — APPROVED** [review]
This feels really good. I think the additional platform testing re: my comment on ODR+exec does not have to hold up merging this PR, since we are reasonably confident in our ODR system+tests.
It looks like you addressed @mitchellh's comments well also. 🎉

File: builtin/aws/lambda/platform_logs.go
Comment (mitchellh) [inline builtin/aws/lambda/platform_logs.go:15]: Nit: this should go into the new line separated section above this because its not a waypoint import.

**Review by mitchellh — APPROVED** [review]
Didn’t run it but looks great.

**Pull request merged**
